src/test/scala/chiselTests/EndToEndSMTBaseSpec.scala

```diff
@@ -0,0 +1,97 @@
+package chiselTests
+
+import chisel3.Module
+import chisel3.stage.{ChiselGeneratorAnnotation, ChiselStage}
+import firrtl.annotations.Annotation
+import firrtl.options.{OutputAnnotationFileAnnotation, TargetDirAnnotation}
+import firrtl.stage.OutputFileAnnotation
+import firrtl.util.BackendCompilationUtilities.timeStamp
+import logger.{LazyLogging, LogLevel, LogLevelAnnotation}
+import org.scalatest.flatspec.AnyFlatSpec
+import os._
+
+abstract class EndToEndSMTBaseSpec extends AnyFlatSpec {
+  def success = MCSuccess
+
+  def fail(k: Int) = MCFail(k)
+
+  def test(dut: () => Module, name: String, expected: MCResult, kmax: Int = 0, annos: Seq[Annotation] = Seq()): Unit = {
+    expected match {
+      case MCFail(k) =>
+        assert(kmax >= k, s"Please set a kmax that includes the expected failing step! ($kmax < $expected)")
+      case _ =>
+    }
+    // @todo rewrite BackendCompilationUtilities
+    val testBaseDir = os.pwd / "test_run_dir" / name
+    os.makeDir.all(testBaseDir)
+    val testDir = os.temp.dir(testBaseDir, timeStamp, deleteOnExit = false)
+    val res = (new ChiselStage).execute(
+      Array("-E", "experimental-smt2"),
+      Seq(
+        LogLevelAnnotation(LogLevel.Error), // silence warnings for tests
+        ChiselGeneratorAnnotation(dut),
+        TargetDirAnnotation(testDir.toString)
+      ) ++ annos
+    )
+    val top = res.collectFirst{case OutputAnnotationFileAnnotation(top) => top}.get
+    assert(res.collectFirst { case _: OutputFileAnnotation => true }.isDefined)
+    val r = Z3ModelChecker.bmc(testDir, top, kmax)
+    assert(r == expected)
+  }
+}
+
+private object Z3ModelChecker extends LazyLogging {
+  def bmc(testDir: Path, main: String, kmax: Int): MCResult = {
+    assert(kmax >= 0 && kmax < 50, "Trying to keep kmax in a reasonable range.")
+    Seq.tabulate(kmax + 1) { k =>
+      val stepFile = testDir / s"${main}_step$k.smt2"
+      os.copy(testDir / s"$main.smt2", stepFile)
+      os.write.append(stepFile,
+        s"""${step(main, k)}
+           |(check-sat)
+           |""".stripMargin)
+      if (!executeStep(stepFile)) return MCFail(k)
```

Comment:
Please add the changes you made to the version from firrtl.

Reply:
Sure:)